R/backimputation.R

```diff
@@ -0,0 +1,86 @@
+#' Impute unobserved generations of infection
+#'
+#' This function imputes incidence prior the first date of reported cases to
+#' address early bias in R_t estimates.
+#' A simple linear model is fitted on shifted, logged-incidence cases, based on
+#' an initial observation window.
+#' Currently, no cases are assumed to be imported.
+#'
+#' @param incid the raw, reported incidence cases.
+#' @param window_b length of the observation window to fit the exponential
+#' growth model for back-imputation
+#'
+#' @return an incidence data.frame, combining back-imputed cases for 100 time
+#' points(with rows indexed by a negative integer rowname) and cases (with rows
+#' indexed by a non-negative integer)
+#' @export
+#'
+#' @examples
+#' incid_all <- ceiling(exp(.3 * 0:20))
+#' incid_trunc <- tail(incid_all, 10)
+#' x <- backimpute_I(incid=incid_trunc, window_b=6)
+#' idx <- as.integer(rownames(x)) > -10
+#' x[idx, ]$local
+#' incid_all
+#'
+#' @importFrom stats lm predict.lm
+
+backimpute_I <- function(incid, window_b) {
+
+    if (inherits(incid, "incidence")) {
+      msg <- "incidence objects are currently not supported by backimpute_I()."
+      stop(msg)
+    }
+
+    stopifnot("Backimputation window needs to contain at least 2 timepoints" =
+        window_b >= 2)
+    stopifnot("Backimputation window needs to have integer length" =
+        window_b %% 1 == 0)
+    stopifnot("Backimputation window should be shorter than observed incidence" =
+        length(incid) >= window_b)
+
+
+    if (window_b <= 5) {
+        msg <- "The backimputation window is short and may lead to an inaccurate estimate of the growth rate."
+        warning(msg)
+    }
+
+    # process observed incidence
+    incid_processed <- process_I(incid)
+    incid_processed[1, ] <- c(sum(incid_processed[1, ]), 0)
+
+    # some cases may be 0, implying -infinite logs
+    safe_shift <- .5
+
+    # backimpute unobserved, previous cases based on first window_b of observations
+    log_incid_start <- data.frame(
+        t = seq(window_b),
+        li = log(incid[1:window_b] + safe_shift)
+   )
+    imputed_t <- seq(from = -100, to = 0)
```

Comment:
@sangeetabhatia03  and @abriz97  I think it would be best to not barcode -100? Or do we think it's ok?

Reply:
Hi @annecori, 

This is a good point that deserves discussion. My answer below covers both this and the `safe_shift` question, as your concerns are similar.

In my upcoming pushed commit, I mention the two decisions in the documentation of `backimpute_I()`:

```{R backimpute_I_documentation}
#' Log-incidence is computed as log(local + 0.5) to avoid -infinite logs.
...
#' @return an incidence data.frame, combining back-imputed cases for a maximum 
#' of 100 time points (with rows indexed by a negative integer rowname) and 
#' cases (with rows #' indexed by a non-negative integer)
```

In both cases, there is no mathematical argument justifying the exact choice of `safe_shift=0.5` or the maximum of 100 time units for which the back-imputation is performed. The heuristic justifications are: 
* I chose `safe_shift=0.5` as an "average" between one and zero cases. This would be a "middle ground" between a world where the observed incidence is correct, and a world in which 1 case has been missed at every time unit. I am happy to consider other heuristics.
* Concerning the maximum of 100 time units for backimputation, this should be an upper bound that should work for most cases. There are two reasons. Firstly, if $M$ is the maximum possible generation interval (GI), the $R_{t}$ estimates are not affected by the incidence at time $t-M$. I would assume that for such delays, the GI pmf $\approx 0$  for most cases of interest (unless we specify generation interval distributions at the hour level). Secondly, when back-imputation is performed, the number of imputed cases will decrease exponentially further into the past. So even a growth rate of (1-p/100) implies that the number of cases at time -100 is $(1 - p/100)^{100} \approx e^{-p}$ the number of cases at time $0$. For p=4, this means that the imputed cases at time -100 are about $2\%$ of the cases at time 0.